File: android/truepad-core/src/main/kotlin/dev/systemslibrarian/truepad/core/Partition.kt
package dev.systemslibrarian.truepad.core

/*
 * TruePad v2 source-material partition — byte-exact twin of src/core/partition2.ts.
 *
 * Combine declared sources by bytewise XOR and carve the combined material into
 * the four secret slices [abEnc E][abAuth 32N][baEnc E][baAuth 32N], with
 * L = 2·(E + 32·N) the length every source must supply. The XOR and this
 * partition are the only operations between sources and secret body — NO KDF,
 * extractor, hash conditioner, or content-based rejection, ever (§7).
 */

private const val KEY_BYTES = 16

// The reference implementation is JavaScript, so every count it can represent
// exactly is bounded by Number.MAX_SAFE_INTEGER. Staying inside the reference's
// domain is what keeps the two implementations in agreement.
private const val MAX_SAFE_INTEGER = 9_007_199_254_740_991L

/** L = 2·(E + 32·N): the exact byte count every declared source must supply. */
fun requiredSourceLength(capacity: Long, capacityRecords: Long): Long {
    require(capacity in 0..MAX_SAFE_INTEGER) { "capacity must be a non-negative safe integer, not $capacity" }
    require(capacityRecords in 0..MAX_SAFE_INTEGER) {
        "capacityRecords must be a non-negative safe integer, not $capacityRecords"
    }
    // The reference computes in Number and is bounded by Number.isSafeInteger.
    // A Kotlin Long reaches further and would wrap silently past where the
    // reference stops being exact, so the ceiling is enforced on the RESULT too.
    val authBytes = AUTH_RECORD_BYTES.toLong() * capacityRecords
    require(authBytes <= MAX_SAFE_INTEGER) { "32 * $capacityRecords exceeds the safe-integer range" }
    val total = 2 * (capacity + authBytes)
    require(total in 0..MAX_SAFE_INTEGER) {
        "2*(E + 32*N) = $total exceeds the safe-integer range for E=$capacity, N=$capacityRecords"
    }
    return total
}

class PairSlices(
    val abEncryption: ByteArray,
    val abAuthentication: ByteArray,
    val baEncryption: ByteArray,
    val baAuthentication: ByteArray,
)

/**
 * Bytewise XOR of the first `length` bytes of every source. All-or-nothing: a
 * source shorter than `length` (or no sources) throws before any byte is
 * combined. Bytes beyond `length` are not read. NEVER inspects or conditions on
 * content — an all-zero combined result is a legitimate draw.
 */
fun combineSources(sources: List<ByteArray>, length: Int): ByteArray {
    require(length >= 0) { "length must be non-negative, not $length" }
    require(sources.isNotEmpty()) { "combineSources needs at least one source" }
    for (i in sources.indices) {
        require(sources[i].size >= length) {
            "source $i supplies ${sources[i].size} bytes but $length are required"
        }
    }
    val combined = ByteArray(length)
    for (source in sources) {
        for (i in 0 until length) combined[i] = (combined[i].toInt() xor source[i].toInt()).toByte()
    }
    return combined
}

/** The §7 partition, exactly. Returns COPIES so a caller can zero `combined`. */
fun partition(combined: ByteArray, capacity: Int, capacityRecords: Int): PairSlices {
    val length = requiredSourceLength(capacity.toLong(), capacityRecords.toLong()).toInt()
    require(combined.size == length) {
        "combined material is ${combined.size} bytes but the partition needs exactly $length " +
            "(2*($capacity + 32*$capacityRecords))"
    }
    val authBytes = AUTH_RECORD_BYTES * capacityRecords
    var cursor = 0
    fun take(count: Int): ByteArray {
        val copy = combined.copyOfRange(cursor, cursor + count)
        cursor += count
        return copy
    }
    return PairSlices(
        abEncryption = take(capacity),
        abAuthentication = take(authBytes),
        baEncryption = take(capacity),
        baAuthentication = take(authBytes),
    )
}

/** Auth record `sequence`: K_s = [32s,32s+16), R_s = [32s+16,32s+32) (slice-local). */
fun authRecordAt(authSlice: ByteArray, sequence: Int): Pair<ByteArray, ByteArray> {
    require(sequence >= 0) { "sequence must be non-negative, not $sequence" }
    // Computed in Long, DELIBERATELY. In Int, sequence * 32 wraps at 2^27: the
    // bound check below would then pass and this would hand back auth record 0's
    // K and R for a completely different sequence - the same one-time key used
    // twice, which is the one failure this product exists to prevent. The
    // reference computes in Number and simply throws. Never narrow this.
    val start = sequence.toLong() * AUTH_RECORD_BYTES
    require(start + AUTH_RECORD_BYTES <= authSlice.size.toLong()) {
        "auth record $sequence needs slice bytes [$start, ${start + AUTH_RECORD_BYTES}) but the slice holds ${authSlice.size}"
    }
    val from = start.toInt()
    val key = authSlice.copyOfRange(from, from + KEY_BYTES)
    val mask = authSlice.copyOfRange(from + KEY_BYTES, from + AUTH_RECORD_BYTES)
    return key to mask
}

File: ios/TruePadKit/Sources/TruePadCore/Partition.swift
/* ============================================================================
 * TruePad v2 source-material partition (FORMAT-V2.md §7, §1.2)
 * ----------------------------------------------------------------------------
 * Byte-exact twin of src/core/partition2.ts. Pure byte movement, nothing else.
 *
 * This module combines declared sources by bytewise XOR and carves the combined
 * material into the four secret slices of a pair:
 *
 *   M[0        .. E)          A->B encryption slice  (byte e ↦ offset e)
 *   M[E        .. E+32N)      A->B authentication slice
 *   M[E+32N    .. 2E+32N)     B->A encryption slice  (byte E+32N+e ↦ offset e)
 *   M[2E+32N   .. 2E+64N)     B->A authentication slice
 *
 * with E = capacity, N = capacityRecords, and L = 2·(E + 32·N) the length every
 * declared source must supply. Every combined byte lands in exactly one slice at
 * exactly one position; the XOR and this partition are the ONLY operations
 * between declared sources and secret body — no KDF, no extractor, no hash
 * conditioner, ever (§7). Within an authentication slice, record s is bytes
 * [32s, 32s+16) as K_s and [32s+16, 32s+32) as R_s (§1.2) — key first, then mask.
 *
 * What this module is NOT: it does not read files, judge uniformity, or enforce
 * the one-file-one-source rule. The verdict for combined material belongs to the
 * generation layer ("Uniform if at least one declared source was uniform and
 * independent of the others").
 *
 * It also never zeroizes: `partition` returns fresh copies precisely so the
 * caller can zero the combined buffer, and each slice, on its own schedule
 * without the copies aliasing it. (Swift arrays are value types, so the copies
 * are structural rather than a discipline anyone has to remember.)
 * ========================================================================= */

public enum Partition {
    /// K_s is the first 16 bytes of a 32-byte auth record; R_s the second 16.
    static let keyBytes = 16

    public enum Failure: Error, Equatable {
        case negativeBudget(String, Int)
        case noSources
        case sourceTooShort(index: Int, supplied: Int, required: Int)
        case combinedWrongLength(supplied: Int, required: Int)
        case authRecordOutOfRange(sequence: Int, sliceCount: Int)
    }

    static func assertBudget(_ name: String, _ value: Int) throws {
        guard value >= 0 else { throw Failure.negativeBudget(name, value) }
    }

    /// L = 2·(E + 32·N): the exact byte count every declared source must supply
    /// (§7). Surplus beyond L is never used.
    public static func requiredSourceLength(capacity: Int, capacityRecords: Int) throws -> Int {
        try assertBudget("capacity", capacity)
        try assertBudget("capacityRecords", capacityRecords)
        return 2 * (capacity + WcOneTime.authRecordBytes * capacityRecords)
    }

    public struct PairSlices: Sendable {
        public let abEncryption: [UInt8]
        public let abAuthentication: [UInt8]
        public let baEncryption: [UInt8]
        public let baAuthentication: [UInt8]
    }

    /// Bytewise XOR of the first `length` bytes of every source. All-or-nothing:
    /// a source shorter than `length` — or no sources at all — throws before any
    /// byte is combined. Bytes beyond `length` are not read.
    public static func combineSources(_ sources: [[UInt8]], length: Int) throws -> [UInt8] {
        try assertBudget("length", length)
        guard !sources.isEmpty else { throw Failure.noSources }
        for (i, source) in sources.enumerated() where source.count < length {
            throw Failure.sourceTooShort(index: i, supplied: source.count, required: length)
        }
        var combined = [UInt8](repeating: 0, count: length)
        for source in sources {
            for i in 0..<length { combined[i] ^= source[i] }
        }
        return combined
    }

    /// The §7 partition, exactly: [abEnc E][abAuth 32N][baEnc E][baAuth 32N].
    /// `combined` must be exactly L = 2·(E + 32·N) bytes.
    public static func partition(_ combined: [UInt8],
                                 capacity: Int,
                                 capacityRecords: Int) throws -> PairSlices {
        let length = try requiredSourceLength(capacity: capacity, capacityRecords: capacityRecords)
        guard combined.count == length else {
            throw Failure.combinedWrongLength(supplied: combined.count, required: length)
        }
        let authBytes = WcOneTime.authRecordBytes * capacityRecords
        var cursor = 0
        func take(_ count: Int) -> [UInt8] {
            let out = Array(combined[cursor..<(cursor + count)])
            cursor += count
            return out
        }
        return PairSlices(
            abEncryption: take(capacity),
            abAuthentication: take(authBytes),
            baEncryption: take(capacity),
            baAuthentication: take(authBytes)
        )
    }

    public struct AuthRecord: Sendable, Equatable {
        public let key: [UInt8]
        public let mask: [UInt8]
    }

    /// Auth record `sequence` out of a direction's authentication slice: bytes
    /// [32s, 32s+16) as the hash key K_s, [32s+16, 32s+32) as the mask R_s —
    /// slice-local offsets (§1.2). A sequence past the slice throws: no record
    /// beyond capacityRecords exists, and this module never invents one.
    public static func authRecord(in authSlice: [UInt8], sequence: Int) throws -> AuthRecord {
        try assertBudget("sequence", sequence)
        let start = sequence * WcOneTime.authRecordBytes
        guard start + WcOneTime.authRecordBytes <= authSlice.count else {
            throw Failure.authRecordOutOfRange(sequence: sequence, sliceCount: authSlice.count)
        }
        return AuthRecord(
            key: Array(authSlice[start..<(start + keyBytes)]),
            mask: Array(authSlice[(start + keyBytes)..<(start + WcOneTime.authRecordBytes)])
        )
    }
}
